blockchain/stakeversion.go

```diff
@@ -0,0 +1,361 @@
+	// Walk blockchain backwards to find interval.
+	node, err := b.findStakeVersionPriorNode(prevNode)
+	if err != nil {
+		return 0, nil
+	}
+
+	// Iterate over versions until we find a majority.
+	iterNode := node
+	for iterNode != nil {
+		version, err := b.calcVoterVersionInterval(iterNode)
+		if err == nil {
+			return version, iterNode
+		}
+		if err != errVoterVersionMajorityNotFound {
+			break
+		}
+
+		// findStakeVersionPriorNode increases the height so we need to
+		// compensate by loading the prior node.
+		iterNode, err = b.getPrevNodeFromNode(iterNode)
+		if err != nil {
+			break
+		}
+
+		// Walk blockchain back to prior interval.
+		iterNode, err = b.findStakeVersionPriorNode(iterNode)
+		if err != nil {
+			break
+		}
+	}
+
+	// We didn't find a marority version.
+	return 0, nil
+}
+
+// calcStakeVersion calculates the header stake version based on voter
+// versions.  If there is a majority of voter versions it uses the header stake
+// version to prevent reverting to a prior version.
+//
+// This function MUST be called with the chain state lock held (for writes).
+func (b *BlockChain) calcStakeVersion(prevNode *blockNode) uint32 {
+	version, node := b.calcVoterVersion(prevNode)
+	if version == 0 || node == nil {
+		// short circuit
+		return 0
+	}
+
+	// Walk chain backwards to start of node interval (start of current
+	// period) Note that calcWantHeight returns the LAST height of the
+	// prior interval; hence the + 1.
+	startIntervalHeight := calcWantHeight(b.chainParams.StakeValidationHeight,
+		b.chainParams.StakeVersionInterval, node.height) + 1
+	iterNode := node
+	for iterNode.height > startIntervalHeight {
+		var err error
+		iterNode, err = b.getPrevNodeFromNode(iterNode)
+		if err != nil || iterNode == nil {
+			return 0
+		}
+	}
+
+	// See if we are enforcing V3 blocks yet.  Just return V0 since it it
+	// wasn't enforced and therefore irrelevant.
+	if !b.isMajorityVersion(3, iterNode,
+		b.chainParams.BlockRejectNumRequired) {
+		return 0
+	}
+
+	if b.isStakeMajorityVersion(version, node) {
+		priorVersion, _ := b.calcPriorStakeVersion(node)
+		if version > priorVersion {
+			return version
+		} else {
+			return priorVersion
+		}
+	}
+
+	return version
+}
+
+// calcStakeVersionByHash calculates the last prior valid majority stake
+// version.  If the current interval does not have a majority stake version
+// it'll go back to the prior interval.  It'll keep going back up to the
+// minimum height at which point we know the version was 0.
+//
+// This function MUST be called with the chain state lock held (for writes).
+func (b *BlockChain) calcStakeVersionByHash(hash *chainhash.Hash) (uint32, error) {
+	prevNode, err := b.findNode(hash, 0)
+	if err != nil {
+		return 0, err
+	}
+
+	return b.calcStakeVersionByNode(prevNode)
+}
+
+// calcStakeVersionByNode is identical to calcStakeVersionByHash but takes a
+// *blockNode instead.
+func (b *BlockChain) calcStakeVersionByNode(prevNode *blockNode) (uint32, error) {
```

Comment:
What's the reasoning for this function?  I would suggest removing it since the code above can simply do:

```Go
return b.calcStakeVersion(prevNode), nil
```

Also, if the function is kept, it would need to have the locking requirement called out:

```Go
+//
+// This function MUST be called with the chain state lock held (for writes).
```